File: src/models/link.rs
use std::net::IpAddr;

#[derive(Debug, Clone, PartialEq)]
pub struct Link {
    pub source_ip: Option<IpAddr>,
    pub source_chassis_id: Option<String>,
    pub source_interface: Option<String>,

    pub target_ip: Option<IpAddr>,
    pub target_chassis_id: Option<String>,
    pub target_hostname: Option<String>,
    pub target_port_id: String,
    pub target_port_description: Option<String>,
}

File: oxian_py/models/link.py
from __future__ import annotations

from ipaddress import IPv4Address, IPv6Address
from pydantic import BaseModel, ConfigDict


class Link(BaseModel):
    """Represents a topological link between two network devices or interfaces."""

    model_config = ConfigDict(arbitrary_types_allowed=True, use_enum_values=True)

    source_ip: str | IPv4Address | IPv6Address | None = None
    source_chassis_id: str | None = None
    source_interface: str | None = None

    target_ip: str | IPv4Address | IPv6Address | None = None
    target_chassis_id: str | None = None
    target_hostname: str | None = None
    target_port_id: str = ""
    target_port_description: str | None = None
